nighthawk/source/client/benchmark_client_impl.cc

```diff
@@ -0,0 +1,226 @@
+                                      *transport_socket.mutable_config());
+    }
+
+    // TODO(oschaaf): Ideally we'd just re-use Tls::Upstream::createTransportFactory().
+    // But instead of doing that, we need to perform some of what that implements ourselves here,
+    // so we can skip message validation which may trigger an assert in integration tests.
+    auto& config_factory = Envoy::Config::Utility::getAndCheckFactory<
+        Envoy::Server::Configuration::UpstreamTransportSocketConfigFactory>(
+        transport_socket.name());
+    Envoy::ProtobufTypes::MessagePtr message =
+        Envoy::Config::Utility::translateToFactoryConfig(transport_socket, config_factory);
+
+    ssl_context_manager_.reset(
+        new Envoy::Extensions::TransportSockets::Tls::ContextManagerImpl(api_.timeSource()));
+    transport_socket_factory_context_ = std::make_unique<Ssl::MinimalTransportSocketFactoryContext>(
+        store_->createScope("client."), dispatcher_, generator_, *store_, api_,
+        *ssl_context_manager_);
+
+    auto client_config =
+        std::make_unique<Envoy::Extensions::TransportSockets::Tls::ClientContextConfigImpl>(
+            dynamic_cast<const envoy::api::v2::auth::UpstreamTlsContext&>(*message),
+            *transport_socket_factory_context_);
+    socket_factory =
+        std::make_unique<Envoy::Extensions::TransportSockets::Tls::ClientSslSocketFactory>(
+            std::move(client_config), transport_socket_factory_context_->sslContextManager(),
+            transport_socket_factory_context_->statsScope());
+  } else {
+    socket_factory = std::make_unique<Envoy::Network::RawBufferSocketFactory>();
+  };
+
+  cluster_ = std::make_unique<Envoy::Upstream::ClusterInfoImpl>(
+      cluster_config, bind_config, runtime, std::move(socket_factory),
+      store_->createScope("client."), false /*added_via_api*/);
+
+  auto host = std::shared_ptr<Envoy::Upstream::Host>{new Envoy::Upstream::HostImpl(
+      cluster_, uri_->host_and_port(), target_address_,
+      envoy::api::v2::core::Metadata::default_instance(), 1 /* weight */,
+      envoy::api::v2::core::Locality(),
+      envoy::api::v2::endpoint::Endpoint::HealthCheckConfig::default_instance(), 0,
+      envoy::api::v2::core::HealthStatus::HEALTHY)};
+
+  Envoy::Network::ConnectionSocket::OptionsSharedPtr options =
+      std::make_shared<Envoy::Network::ConnectionSocket::Options>();
+  if (use_h2_) {
+    pool_ = std::make_unique<Envoy::Http::Http2::ProdConnPoolImpl>(
+        dispatcher_, host, Envoy::Upstream::ResourcePriority::Default, options);
+  } else {
+    pool_ = std::make_unique<Envoy::Http::Http1::ProdConnPoolImpl>(
+        dispatcher_, host, Envoy::Upstream::ResourcePriority::Default, options);
+  }
+  return true;
+}
+
+void BenchmarkClientHttpImpl::terminate() { pool_.reset(); }
+
+StatisticPtrMap BenchmarkClientHttpImpl::statistics() const {
+  StatisticPtrMap statistics;
+  statistics[connect_statistic_->id()] = connect_statistic_.get();
```

Comment:
Just a placeholder comment re: our discussion on filing an issue on better integration with Envoy stats.

Reply:
Added to the track-for-future issue